File: 0098-Validate-Binary-Search-Tree/0098.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

//  Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();

class Solution 
{
public:
    bool isValidBST(TreeNode* root) 
    {
        return _isVaildBST(root, nullptr, nullptr);
    }
private:
    bool _isVaildBST(TreeNode* root, TreeNode* left, TreeNode* right)
    {
        if (root == nullptr) return true;
        if (left and left->val >= root->val) return false;
        if (right and right->val <= root->val) return false;
        return _isVaildBST(root->left, left, root) and _isVaildBST(root->right, root, right);
    }
};

File: src/0098-Validate-Binary-Search-Tree/0098.py
# class Solution:
#     def _isValidBST(self, root, left=None, right=None):
#         if not root:
#             return True

#         if left and left.val >= root.val:
#             return False

#         if right and right.val <= root.val:
#             return False

#         return self._isValidBST(root.left, left, root) and self._isValidBST(root.right, root, right)

#     def isValidBST(self, root):
#         """
#         :type root: TreeNode
#         :rtype: bool
#         """
#         return self._isValidBST(root)


class Solution:
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True

        pre = None
        stack = list()
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                root = stack.pop()
                if pre and root.val <= pre.val:
                        return False
                
                pre = root
                root = root.right

        return True